perfschema/statement.go

```diff
@@ -0,0 +1,299 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package perfschema
+
+import (
+	"fmt"
+	"reflect"
+	"runtime"
+	"sync/atomic"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+// statementInfo defines statement instrument information.
+type statementInfo struct {
+	// The registered statement key
+	key uint64
+	// The name of the statement instrument to register
+	name string
+}
+
+// StatementState provides temporary storage to a statement runtime statistics.
+// TODO:
+// 1. support statement digest.
+// 2. support prepared statement.
+type StatementState struct {
+	// Connection identifier
+	connID uint64
+	// Statement information
+	info *statementInfo
+	// Statement type
+	stmtType reflect.Type
+	// Source file and line number
+	source string
+	// Timer name
+	timerName enumTimerName
+	// Timer start
+	timerStart int64
+	// Timer end
+	timerEnd int64
+	// Locked time
+	lockTime int64
+	// SQL statement string
+	sqlText string
+	// Current schema name
+	schemaName string
+	// Number of errors
+	errNum uint32
+	// Number of warnings
+	warnNum uint32
+	// Rows affected
+	rowsAffected uint64
+	// Rows sent
+	rowsSent uint64
+	// Rows examined
+	rowsExamined uint64
+	// Metric, temporary tables created on disk
+	createdTmpDiskTables uint32
+	// Metric, temproray tables created
+	createdTmpTables uint32
+	// Metric, number of select full join
+	selectFullJoin uint32
+	// Metric, number of select full range join
+	selectFullRangeJoin uint32
+	// Metric, number of select range
+	selectRange uint32
+	// Metric, number of select range check
+	selectRangeCheck uint32
+	// Metric, number of select scan
+	selectScan uint32
+	// Metric, number of sort merge passes
+	sortMergePasses uint32
+	// Metric, number of sort merge
+	sortRange uint32
+	// Metric, number of sort rows
+	sortRows uint32
+	// Metric, number of sort scans
+	sortScan uint32
+	// Metric, no index used flag
+	noIndexUsed uint8
+	// Metric, no good index used flag
+	noGoodIndexUsed uint8
+}
+
+const (
+	// Maximum allowed number of elements in table events_statements_history.
+	// TODO: make it configurable?
+	stmtsHistoryElemMax uint64 = 1024
+)
+
+var (
+	stmtInfos = make(map[reflect.Type]*statementInfo)
+)
+
+func (ps *perfSchema) RegisterStatement(category, name string, elem interface{}) {
+	instrumentName := fmt.Sprintf("%s%s/%s", statementInstrumentPrefix, category, name)
+	key, err := ps.addInstrument(instrumentName)
+	if err != nil {
+		// just ignore, do nothing else.
+		return
+	}
+
+	stmtInfos[reflect.TypeOf(elem)] = &statementInfo{
+		key:  key,
+		name: instrumentName,
+	}
+}
+
+func (ps *perfSchema) StartStatement(sql string, connID uint64, elem interface{}) *StatementState {
+	stmtType := reflect.TypeOf(elem)
```

Comment:
After discussed with @coocood , I will:
1) Keep TypeOf (about 27ns per call), it is acceptable so far
2) Optimize runtime.Caller (about 563ns per call), not acceptable